Revamp handling of notificationclick and notificationclose

beidson opened the pull request:
#### 135df2fc81755218135e1a49886e6efe74a30a99
<pre>
Revamp handling of notificationclick and notificationclose
<a href="https://rdar.apple.com/131365403">rdar://131365403</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=278825">https://bugs.webkit.org/show_bug.cgi?id=278825</a>

Reviewed by Ben Nham.

This reworks some logic in webpushd and in the WebKit UIProcess to properly
capture and route &quot;notificationclick&quot; and &quot;notificationclose&quot; events for
clients using built-in notifications.

* Source/WebKit/Platform/spi/ios/BaseBoardSPI.h:
* Source/WebKit/Platform/spi/ios/UIKitSPI.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
(-[WKWebsiteDataStore _processWebCorePersistentNotificationClick:completionHandler:]):
(-[WKWebsiteDataStore _processPersistentNotificationClick:completionHandler:]):
(-[WKWebsiteDataStore _processWebCorePersistentNotificationClose:completionHandler:]):
(-[WKWebsiteDataStore _processPersistentNotificationClose:completionHandler:]):
(+[WKWebsiteDataStore _setWebPushActionHandler:]):
(-[WKWebsiteDataStore _handleNextPushMessage]):
(-[WKWebsiteDataStore _handleWebPushAction:]):
(-[_WKWebsiteDataStoreBSActionHandler _respondToApplicationActions:fromTransitionContext:]):
* Source/WebKit/UIProcess/API/Cocoa/_WKWebPushAction.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebPushAction.mm:
(+[_WKWebPushAction _webPushActionWithNotificationResponse:]):
* Source/WebKit/UIProcess/API/Cocoa/_WKWebPushActionInternal.h: Copied from Source/WebKit/UIProcess/API/Cocoa/_WKWebPushAction.h.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/webpushd/WebPushDaemon.mm:
(WebPushD::platformDefaultActionBundleIdentifier):
(WebPushD::WebPushDaemon::notifyClientPushMessageIsAvailable):
(WebPushD::WebPushDaemon::showNotification):

Canonical link: <a href="https://commits.webkit.org/282885@main">https://commits.webkit.org/282885@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c436688630b7a71dc7ff30556d5daebe65e3ec8f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/64577 "5 style errors") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/43942 "Build is in progress. Recent messages:OS: Sonoma (14.5), Xcode: 15.4; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/17172 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/68598 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/15183 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/51700 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/15462 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/51944 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/10472 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/67643 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/40654 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/55878 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/32568 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/37319 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/14056 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/59249 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/13586 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/70297 "Build is in progress. Recent messages:Printed configuration; Running apply-patch; Checked out pull request; Running compile-webkit") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/8522 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/13087 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/59271 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/8556 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/55966 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/59445 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14238 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/7042 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/716 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/39753 "Built successfully") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/40831 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/42014 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/40575 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

bnham commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
```diff
@@ -1345,6 +1389,34 @@ - (void)setWebPushActionHandler:(WKWebsiteDataStore *(^)(_WKWebPushAction *))han
+        if (action.UIActionType == UIActionTypeNotificationResponse) {
+            // Try to pull out a notification response from the action
+            UNNotificationResponse *response = nil;
+            SEL responseSelector = NSSelectorFromString(@"response");
```
Should this be declared in a category method instead so you don't have to `performSelector:`?

beidson replied:
Hmmmm okay.

bnham commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
```diff
@@ -1345,6 +1389,34 @@ - (void)setWebPushActionHandler:(WKWebsiteDataStore *(^)(_WKWebPushAction *))han
+
+            RetainPtr<_WKWebPushAction> webPushAction = [_WKWebPushAction _webPushActionWithNotificationResponse:response];
+            if (!webPushAction) {
+                [unhandled addObject:action];
```
So I presume the way this works is that if `response` is a normal notification and not a web push related one, then `-_webPushActionWithNotificationResponse:` returns nil, and adding the action to the unhandled set lets the action bubble down to a lower priority action handler?

beidson replied:
Exactly!

bnham commented on Source/WebKit/WebKit.xcodeproj/project.pbxproj:
```diff
@@ -1960,7 +1961,6 @@
 		A14F9B762B68CA6C00AD9C56 /* WKSLinearMediaPlayer.h in Headers */ = {isa = PBXBuildFile; fileRef = A14F9B742B68CA6B00AD9C56 /* WKSLinearMediaPlayer.h */; settings = {ATTRIBUTES = (Private, ); }; };
 		A14F9B772B68CA6C00AD9C56 /* LinearMediaPlayer.swift in Sources */ = {isa = PBXBuildFile; fileRef = A14F9B752B68CA6B00AD9C56 /* LinearMediaPlayer.swift */; };
 		A15EEDE61E301CEE000069B0 /* WKPasswordView.h in Headers */ = {isa = PBXBuildFile; fileRef = A15EEDE41E301CEE000069B0 /* WKPasswordView.h */; };
-		A16E66752581A44800EE1749 /* WebKit.framework in Frameworks */ = {isa = PBXBuildFile; fileRef = 8DC2EF5B0486A6940098B216 /* WebKit.framework */; };
```
Is this an intended change? (Same for the other removals.)

beidson replied:
This is just Xcode having its way with me.
I tried resetting it and adding the header again and *boom* this.

I'll try to do it manually.

bnham commented on Source/WebKit/webpushd/WebPushDaemon.mm:
```diff
@@ -936,10 +940,13 @@ static bool connectionMatchesPendingPushMessage(const PushClientConnection& conn
         RELEASE_LOG_ERROR(Push, "Failed to instantiate UNUserNotificationCenter center");
 
+    UNNotificationCategory *category = [UNNotificationCategory categoryWithIdentifier:@"webpushdCategory" actions:@[] intentIdentifiers:@[] options:UNNotificationCategoryOptionCustomDismissAction];
+    center.get().notificationCategories = [NSSet setWithObjects:category, nil];
```
`-[NSSet setWithObject:]`

beidson replied:
Hah!

bnham reviewed (approved):
LGTM pending build fixes. A little bit nervous that `_registerInternalBSActionHandler:` could cause some problems down the road, but I guess not as risky if none of the action handlers have changed for a long time.

beidson committed:
Revamp handling of notificationclick and notificationclose
rdar://131365403
https://bugs.webkit.org/show_bug.cgi?id=278825

Reviewed by Ben Nham.

This reworks some logic in webpushd and in the WebKit UIProcess to properly
capture and route "notificationclick" and "notificationclose" events for
clients using built-in notifications.

* Source/WebKit/Platform/spi/ios/BaseBoardSPI.h:
* Source/WebKit/Platform/spi/ios/UIKitSPI.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
(-[WKWebsiteDataStore _processWebCorePersistentNotificationClick:completionHandler:]):
(-[WKWebsiteDataStore _processPersistentNotificationClick:completionHandler:]):
(-[WKWebsiteDataStore _processWebCorePersistentNotificationClose:completionHandler:]):
(-[WKWebsiteDataStore _processPersistentNotificationClose:completionHandler:]):
(+[WKWebsiteDataStore _setWebPushActionHandler:]):
(-[WKWebsiteDataStore _handleNextPushMessage]):
(-[WKWebsiteDataStore _handleWebPushAction:]):
(-[_WKWebsiteDataStoreBSActionHandler _respondToApplicationActions:fromTransitionContext:]):
* Source/WebKit/UIProcess/API/Cocoa/_WKWebPushAction.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebPushAction.mm:
(+[_WKWebPushAction _webPushActionWithNotificationResponse:]):
* Source/WebKit/UIProcess/API/Cocoa/_WKWebPushActionInternal.h: Copied from Source/WebKit/UIProcess/API/Cocoa/_WKWebPushAction.h.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/webpushd/WebPushDaemon.mm:
(WebPushD::platformDefaultActionBundleIdentifier):
(WebPushD::WebPushDaemon::notifyClientPushMessageIsAvailable):
(WebPushD::WebPushDaemon::showNotification):

Canonical link: https://commits.webkit.org/282885@main
